### PR TITLE
Ensure private headers are declared as such in a framework's generated module map.

### PR DESCRIPTION
**Pull request by segiddins**
Closes https://github.com/CocoaPods/CocoaPods/issues/2974.

\c @mrackwitz @kylef

### CODE REVIEW

**Comment by neonichu** [comment]
Do we have a good reason to generate our module map differently from Xcode defaults (see <https://github.com/CocoaPods/CocoaPods/issues/2974#issuecomment-68075930>)?

File: lib/cocoapods/generator/module_map.rb [hunk 1]
```diff
@@ -35,14 +40,24 @@ def save_as(path)
...
+        result << "\n#{generate_private_header_exports}" unless private_headers.empty?
+        result << "}\n"
+      end
+
+      private
+
+      def generate_private_header_exports
+        private_headers.reduce('') do |string, header|
+          string << %(  private header "#{header}\n")
+        end
```
Comment (kylef) [inline lib/cocoapods/generator/module_map.rb]: I'm not entirely sure if this is correct.

Shouldn't the output be something like:

```c
module Foo {
  header "Foo.h"

  explicit module Private {
    header "Foo_Private.h"
  }
}
```

Comment (kylef) [inline lib/cocoapods/generator/module_map.rb]: There doesn't seem to be any tests to ensure `generate` works, can we add some? (I understand if you feel this is outside of this PR, can be done separately, but I think it makes sense to add them in this PR to make sure this works correctly).

Comment (segiddins) [inline lib/cocoapods/generator/module_map.rb]: Sure, can add some tests later. 

-Samuel E. Giddins

On Jan 5, 2015, at 8:21 AM, Kyle Fuller <notifications@github.com> wrote:

In lib/cocoapods/generator/module_map.rb:

>          eos
> +
> +        result << "\n#{generate_private_header_exports}" unless private_headers.empty?
> +        result << "}\n"
> +      end
> +
> +      private
> +
> +      def generate_private_header_exports
> +        private_headers.reduce('') do |string, header|
> +          string << %(  private header "#{header}\n")
> +        end
There doesn't seem to be any tests to ensure generate works, can we add some? (I understand if you feel this is outside of this PR, can be done separately, but I think it makes sense to add them in this PR to make sure this works correctly).

Comment (segiddins) [inline lib/cocoapods/generator/module_map.rb]: > A header with the private specifier may not be included from outside the module itself.

Comment (segiddins) [inline lib/cocoapods/generator/module_map.rb]: > There doesn't seem to be any tests to ensure generate works

That's the integration specs

**Comment by segiddins** [comment]
note to self: add unit test for spec with private headers

**Comment by segiddins** [comment]
@kylef rebased + added spec

**Comment by neonichu** [comment]
Hm, this only works if one specifies `private_header_files` manually, is that intended? I would have thought this would include all private headers, regardless of definition. Moreover, the private headers are not copied to the framework by CP, so this wouldn't even do anything rn, would it?

**Comment by kylef** [comment]
@neonichu A header wouldn't be private unless it has presence in the pod spec's `private_header_files`?

**Comment by neonichu** [comment]
@kylef seems unintuitive for me, especially since we introduced *Pods/Headers/{Private,Public}* - that suggest that a header is either considered private or public by CP and cannot be neither.

**Comment by neonichu** [comment]
To clarify what I mean, the current implementation in this PR would add `b.h` to the module map in this scenario:

```
source_files = a.h b.h
private_header_files = b.h
```

but not in this one:

```
source_files = a.h b.h
public_header_files = a.h
```

**Comment by segiddins** [comment]
Boris, I'm pretty sure it would in the latter. Check out the file accessor. 

-Samuel E. Giddins

On Jan 19, 2015, at 4:02 AM, Boris Bügling <notifications@github.com> wrote:

To clarify what I mean, the current implementation in this PR would add b.h to the module map in this scenario:

source_files = a.h b.h
private_header_files = b.h
but not in this one:

source_files = a.h b.h
public_header_files = a.h

**Comment by neonichu** [comment]
@segiddins didn't work for me in manual testing - will have a look at why that is the case, first wanted to confirm that I am not missing something and that behaviour was intended.

**Comment by neonichu** [comment]
@segiddins So the file accessor is actually the issue.

<https://github.com/CocoaPods/CocoaPods/pull/3004/files#diff-a1bd48eb509ce3365557319fdda59f25R130> contains only the contents of the `private_header_files` attribute of a spec. I think `headers - public_headers` should actually be the implementation of that method.

But fixing that breaks the build of frameworks anyway, possibly because the private headers are not actually copied into the framework. Unfortunately, the error message is only the ever elusive "Could not build Objective-C module".

**Comment by neonichu** [comment]
Not copying the headers is indeed the issue, we need to add handling for private headers here <https://github.com/CocoaPods/CocoaPods/blob/seg-module-map-private-headers/lib/cocoapods/installer/target_installer/pod_target_installer.rb#L68>, similar to the public ones. 

As this copies the headers into the `PrivateHeaders` directory, this also clears up my earlier confusion here <https://github.com/CocoaPods/CocoaPods/pull/3004#issuecomment-68684803>

**Comment by segiddins** [comment]
I think https://github.com/CocoaPods/CocoaPods/pull/3145 will make this work?

**Comment by segiddins** [comment]
@mrackwitz please see that this will work in conjunction with the custom modulemap PR?

**Comment by mrackwitz** [comment]
We could add [`-Wincomplete-umbrella`](http://clang.llvm.org/docs/Modules.html#header-declaration) to all generated Pod targets:
>Any headers not included by the umbrella header should have explicit header declarations. Use the -Wincomplete-umbrella warning option to ask Clang to complain about headers not covered by the umbrella header or the module map.

This could help to find misconfigurations of the header visibility in the podspec with `pod [spec|lib] lint`. This could also be helpful in combination with custom modulemaps or we disable it in that case and don't add this additional flag for warnings. But the documentation doesn't cover how clang handles `project` headers, which are since #3145 possible with CocoaPods, by adding headers in the `source_files` of a podspec, which are neither matched by `public_headers` nor by `private_headers`.

**Comment by segiddins** [comment]
@mrackwitz if we add `-Wincomplete-umbrella`, will this be mergable?

**Comment by mrackwitz** [comment]
@segiddins: That should go in a separate issue.

**Comment by segiddins** [comment]
In that case, what needs to be done for this to be merged?

**Comment by mrackwitz** [comment]
Rebuilding the integration spec. :white_check_mark:

**Pull request merged**